File: packages/core-java/src/main/java/com/openpasskey/core/MLDSA65.java
package com.openpasskey.core;

import org.bouncycastle.pqc.crypto.mldsa.MLDSAParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSAPublicKeyParameters;
import org.bouncycastle.pqc.crypto.mldsa.MLDSASigner;

/**
 * ML-DSA-65 (Dilithium3, FIPS 204) signature verification.
 */
public final class MLDSA65 {
    private MLDSA65() {}

    public static final int PUBLIC_KEY_SIZE = 1952;

    /**
     * Verify an ML-DSA-65 signature.
     * ML-DSA signs the message directly (no additional hashing).
     *
     * @param publicKeyBytes raw ML-DSA-65 public key (1952 bytes)
     * @param verifyData     authData || SHA256(clientDataJSON)
     * @param signature      ML-DSA-65 signature
     * @return true if the signature is valid
     */
    public static boolean verify(byte[] publicKeyBytes, byte[] verifyData, byte[] signature)
            throws WebAuthnException {
        try {
            MLDSAPublicKeyParameters pubKeyParams = new MLDSAPublicKeyParameters(
                    MLDSAParameters.ml_dsa_65, publicKeyBytes);

            MLDSASigner signer = new MLDSASigner();
            signer.init(false, pubKeyParams);
            signer.update(verifyData, 0, verifyData.length);
            return signer.verifySignature(signature);
        } catch (Exception e) {
            return false;
        }
    }
}

File: packages/core-dotnet/MLDSA65.cs
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;
using PeterO.Cbor;

namespace OpenPasskey.Core;

internal static class MLDSA65
{
    public static void Verify(byte[] coseKeyBytes, byte[] verifyData, byte[] signature)
    {
        CBORObject coseKey = CBORObject.DecodeFromBytes(coseKeyBytes);
        int kty = coseKey[CBORObject.FromObject(1)].AsInt32();
        int alg = coseKey[CBORObject.FromObject(3)].AsInt32();

        if (kty != CoseConstants.KtyMLDSA || alg != CoseConstants.AlgMLDSA65)
            throw new WebAuthnException("unsupported_cose_algorithm");

        byte[] pub = coseKey[CBORObject.FromObject(-1)].GetByteString();

        VerifyWithRawKey(pub, verifyData, signature);
    }

    public static void VerifyWithRawKey(byte[] rawPubKeyBytes, byte[] verifyData, byte[] signature)
    {
        var pubKeyParams = MLDsaPublicKeyParameters.FromEncoding(MLDsaParameters.ml_dsa_65, rawPubKeyBytes);
        var signer = new MLDsaSigner(MLDsaParameters.ml_dsa_65, false);
        signer.Init(false, pubKeyParams);
        signer.BlockUpdate(verifyData, 0, verifyData.Length);
        bool valid = signer.VerifySignature(signature);

        if (!valid)
            throw new WebAuthnException("signature_invalid");
    }
}
